fix: state divergente

Error: ELBv2 Load Balancer (aws-app-platform-dev-alb) already exists

  with module.alb.aws_lb.this,
  on ../../modules/alb/main.tf line 29, in resource "aws_lb" "this":
  29: resource "aws_lb" "this" {


Error: ELBv2 Target Group (aws-app-platform-dev-tg) already exists

  with module.alb.aws_lb_target_group.this,
  on ../../modules/alb/main.tf line 41, in resource "aws_lb_target_group" "this":
  41: resource "aws_lb_target_group" "this" {


Error: creating CloudWatch Logs Log Group (/ecs/aws-app-platform-dev): operation error CloudWatch Logs: CreateLogGroup, https response error StatusCode: 400, RequestID: a7d31400-24d9-4d67-9eb0-75a249b01a36, ResourceAlreadyExistsException: The specified log group already exists

  with module.ecs.aws_cloudwatch_log_group.this,
  on ../../modules/ecs/main.tf line 7, in resource "aws_cloudwatch_log_group" "this":
   7: resource "aws_cloudwatch_log_group" "this" {


Error: creating IAM Role (aws-app-platform-dev-ecs-execution-role): operation error IAM: CreateRole, https response error StatusCode: 409, RequestID: d55ac5c3-a31d-48f0-abb8-fbb8006d2480, EntityAlreadyExists: Role with name aws-app-platform-dev-ecs-execution-role already exists.

  with module.ecs.aws_iam_role.execution,
  on ../../modules/ecs/main.tf line 24, in resource "aws_iam_role" "execution":
  24: resource "aws_iam_role" "execution" {


Error: creating IAM Role (aws-app-platform-dev-ecs-task-role): operation error IAM: CreateRole, https response error StatusCode: 409, RequestID: a440e6f7-4c8e-4388-9100-f29b30d8585b, EntityAlreadyExists: Role with name aws-app-platform-dev-ecs-task-role already exists.

  with module.ecs.aws_iam_role.task,
  on ../../modules/ecs/main.tf line 46, in resource "aws_iam_role" "task":
  46: resource "aws_iam_role" "task" {


Error: creating RDS DB Subnet Group (aws-app-platform-dev-db-subnet-group): operation error RDS: CreateDBSubnetGroup, https response error StatusCode: 400, RequestID: a058faef-9062-493d-92d7-d27da824fe7b, DBSubnetGroupAlreadyExists: The DB subnet group 'aws-app-platform-dev-db-subnet-group' already exists.

  with module.rds.aws_db_subnet_group.this,
  on ../../modules/rds/main.tf line 5, in resource "aws_db_subnet_group" "this":
   5: resource "aws_db_subnet_group" "this" {

Error: Process completed with exit code 1.

--- a/.github/workflows/infra-apply.yml
+++ b/.github/workflows/infra-apply.yml
@@ -50,7 +50,7 @@ jobs:
           fi
 
       - name: Terraform init
-        run: terraform init -input=false
+        run: terraform init -input=false -reconfigure
 
       - name: Terraform apply
         run: terraform apply -input=false -auto-approve -no-color
@@ -91,7 +91,7 @@ jobs:
           fi
 
       - name: Terraform init
-        run: terraform init -input=false
+        run: terraform init -input=false -reconfigure
 
       - name: Terraform apply
         run: terraform apply -input=false -auto-approve -no-color

--- a/.github/workflows/infra-plan.yml
+++ b/.github/workflows/infra-plan.yml
@@ -78,7 +78,7 @@ jobs:
           fi
 
       - name: Terraform init
-        run: terraform init -input=false
+        run: terraform init -input=false -reconfigure
 
       - name: Terraform plan (output file only)
         shell: bash

--- a/README.md
+++ b/README.md
@@ -68,6 +68,10 @@ terraform plan
 terraform apply
 ```
 
+Guia de bootstrap do backend remoto:
+
+- `docs/terraform-backend-bootstrap.md`
+
 ## Nota de custo (Free Tier)
 
 Para compatibilidade com contas AWS Free Tier, o RDS esta configurado com `backup_retention_period = 0` nos ambientes `dev` e `prod`.

--- a/docs/terraform-backend-bootstrap.md
+++ b/docs/terraform-backend-bootstrap.md
@@ -1,0 +1,46 @@
+# Terraform Backend Bootstrap
+
+Este projeto usa backend remoto S3 + DynamoDB para evitar divergencia de state entre execucoes locais e GitHub Actions.
+
+## 1) Criar backend remoto (uma vez)
+
+```bash
+cd infra/bootstrap
+terraform init
+terraform apply
+```
+
+Recursos criados:
+
+- Bucket S3 de state com versioning, encryption e bloqueio publico
+- Tabela DynamoDB para lock de state
+
+## 2) Inicializar ambientes com backend remoto
+
+```bash
+terraform -chdir=infra/envs/dev init -reconfigure
+terraform -chdir=infra/envs/prod init -reconfigure
+```
+
+## 3) Se recursos ja existem na AWS (caso de migracao)
+
+Se aparecer erro `already exists`, importe os recursos para o state remoto do ambiente correspondente.
+
+Exemplo (dev):
+
+```bash
+terraform -chdir=infra/envs/dev import module.alb.aws_lb.this <alb_arn>
+terraform -chdir=infra/envs/dev import module.alb.aws_lb_target_group.this <target_group_arn>
+terraform -chdir=infra/envs/dev import module.ecs.aws_cloudwatch_log_group.this /ecs/aws-app-platform-dev
+terraform -chdir=infra/envs/dev import module.ecs.aws_iam_role.execution aws-app-platform-dev-ecs-execution-role
+terraform -chdir=infra/envs/dev import module.ecs.aws_iam_role.task aws-app-platform-dev-ecs-task-role
+terraform -chdir=infra/envs/dev import module.rds.aws_db_subnet_group.this aws-app-platform-dev-db-subnet-group
+```
+
+Depois:
+
+```bash
+terraform -chdir=infra/envs/dev plan
+```
+
+Repita para `prod` quando aplicavel.

--- a/infra/bootstrap/main.tf
+++ b/infra/bootstrap/main.tf
@@ -1,0 +1,64 @@
+terraform {
+  required_version = ">= 1.6.0"
+
+  required_providers {
+    aws = {
+      source  = "hashicorp/aws"
+      version = "~> 5.0"
+    }
+  }
+}
+
+provider "aws" {
+  region = var.aws_region
+}
+
+resource "aws_s3_bucket" "tf_state" {
+  bucket = var.state_bucket_name
+
+  tags = merge(var.tags, {
+    Name = var.state_bucket_name
+  })
+}
+
+resource "aws_s3_bucket_versioning" "tf_state" {
+  bucket = aws_s3_bucket.tf_state.id
+
+  versioning_configuration {
+    status = "Enabled"
+  }
+}
+
+resource "aws_s3_bucket_server_side_encryption_configuration" "tf_state" {
+  bucket = aws_s3_bucket.tf_state.id
+
+  rule {
+    apply_server_side_encryption_by_default {
+      sse_algorithm = "AES256"
+    }
+  }
+}
+
+resource "aws_s3_bucket_public_access_block" "tf_state" {
+  bucket = aws_s3_bucket.tf_state.id
+
+  block_public_acls       = true
+  block_public_policy     = true
+  ignore_public_acls      = true
+  restrict_public_buckets = true
+}
+
+resource "aws_dynamodb_table" "tf_lock" {
+  name         = var.lock_table_name
+  billing_mode = "PAY_PER_REQUEST"
+  hash_key     = "LockID"
+
+  attribute {
+    name = "LockID"
+    type = "S"
+  }
+
+  tags = merge(var.tags, {
+    Name = var.lock_table_name
+  })
+}

--- a/infra/bootstrap/outputs.tf
+++ b/infra/bootstrap/outputs.tf
@@ -1,0 +1,9 @@
+output "state_bucket_name" {
+  description = "Bucket S3 para state remoto"
+  value       = aws_s3_bucket.tf_state.bucket
+}
+
+output "lock_table_name" {
+  description = "Tabela DynamoDB para lock"
+  value       = aws_dynamodb_table.tf_lock.name
+}

--- a/infra/bootstrap/variables.tf
+++ b/infra/bootstrap/variables.tf
@@ -1,0 +1,26 @@
+variable "aws_region" {
+  description = "Regiao AWS para recursos de backend"
+  type        = string
+  default     = "us-east-1"
+}
+
+variable "state_bucket_name" {
+  description = "Nome globalmente unico do bucket de state"
+  type        = string
+  default     = "aws-app-platform-tfstate-229371276586-us-east-1"
+}
+
+variable "lock_table_name" {
+  description = "Nome da tabela DynamoDB para lock de state"
+  type        = string
+  default     = "aws-app-platform-terraform-locks"
+}
+
+variable "tags" {
+  description = "Tags comuns"
+  type        = map(string)
+  default = {
+    project    = "aws-app-platform"
+    managed_by = "terraform"
+  }
+}

--- a/infra/envs/dev/main.tf
+++ b/infra/envs/dev/main.tf
@@ -1,6 +1,14 @@
 terraform {
   required_version = ">= 1.6.0"
 
+  backend "s3" {
+    bucket         = "aws-app-platform-tfstate-229371276586-us-east-1"
+    key            = "aws-app-platform/dev/terraform.tfstate"
+    region         = "us-east-1"
+    dynamodb_table = "aws-app-platform-terraform-locks"
+    encrypt        = true
+  }
+
   required_providers {
     aws = {
       source  = "hashicorp/aws"

--- a/infra/envs/prod/main.tf
+++ b/infra/envs/prod/main.tf
@@ -1,6 +1,14 @@
 terraform {
   required_version = ">= 1.6.0"
 
+  backend "s3" {
+    bucket         = "aws-app-platform-tfstate-229371276586-us-east-1"
+    key            = "aws-app-platform/prod/terraform.tfstate"
+    region         = "us-east-1"
+    dynamodb_table = "aws-app-platform-terraform-locks"
+    encrypt        = true
+  }
+
   required_providers {
     aws = {
       source  = "hashicorp/aws"

